packages/story-editor/src/app/story/effects/useHistoryEntry.js

```diff
@@ -23,28 +23,83 @@ import { useEffect, useRef } from '@web-stories-wp/react';
  * Internal dependencies
  */
 import { useHistory } from '../../history';
+import deleteNestedKeys from '../utils/deleteNestedKeys';
+
+// @todo REMOVE ROTATION ANGLE! For testing only!
+const ELEMENT_PROPS_TO_IGNORE = ['resource.baseColor', 'rotationAngle'];
 
 // Record any change to core variables in history (history will know if it's a replay)
 function useHistoryEntry({ story, current, pages, selection, capabilities }) {
   const {
+    state: { currentEntry },
     actions: { stateToHistory },
   } = useHistory();
 
+  const currentHistoryEntryRef = useRef();
+  useEffect(() => {
+    if (currentEntry) {
+      currentHistoryEntryRef.current = JSON.parse(JSON.stringify(currentEntry));
+    }
+  }, [currentEntry]);
+
   const currentPageIndexRef = useRef();
   const selectedElementIdsRef = useRef();
   useEffect(() => {
     currentPageIndexRef.current = current;
     selectedElementIdsRef.current = selection;
   }, [current, selection]);
 
-  useEffect(() => {
-    stateToHistory({
-      story,
-      current: currentPageIndexRef.current,
-      selection: selectedElementIdsRef.current,
-      pages,
-      capabilities,
+  const deleteKeysFromPages = (list) => {
+    // Create a copy of the list not to influence the original.
+    return JSON.parse(JSON.stringify(list)).map((page) => {
+      const { elements } = page;
+      return {
+        ...page,
+        elements: elements.map((element) =>
+          deleteNestedKeys(element, ELEMENT_PROPS_TO_IGNORE)
+        ),
+      };
     });
+  };
+
+  useEffect(() => {
+    // There are some element properties that should not influence history.
+    // Before adding a new history entry, let's check if the only properties that changed
+    // should not influence history. Then we skip adding an entry.
+    let skipAddingEntry = false;
+    if (currentHistoryEntryRef.current && pages?.length) {
+      // If story / capabilities change, we should always add a new entry.
+      const withoutPages = {
+        story,
+        capabilities,
+      };
+      const onlyPagesChanged = Object.keys(withoutPages).every(
+        (key) =>
+          JSON.stringify(withoutPages[key]) ===
+          JSON.stringify(currentHistoryEntryRef.current[key])
+      );
+      // If only pages have changed, check if relevant properties have changed.
+      if (onlyPagesChanged) {
+        const adjustedPages = deleteKeysFromPages(pages);
+        const adjustedEntryPages = deleteKeysFromPages(
+          currentHistoryEntryRef.current.pages
+        );
```

Comment:
Couldn't we do this before storing the ref? In fact, couldn't we do the stringify too - so the ref is a map from the three properties (`story`, `capabilities`, and `pages`) to JSON strings?

Reply:
We could, however, I was trying to save some time here -- this runs only if `onlyPagesChanged`, meaning that it doesn't have to always run. Thoughts?

Reply:
But isn't that most updates? Are almost all updates to pages in everyday workflow?